vault/src/vaults.rs

```diff
@@ -68,12 +68,34 @@ impl<P: RelayPallet + BtcRelayPallet, B: BitcoinCoreApi + Clone> VaultTheftMonit
         }
     }
 
-    async fn report_invalid(&self, vault_id: &AccountId, proof: &[u8], raw_tx: &[u8]) -> Result<(), Error> {
+    async fn report_invalid(
+        &self,
+        vault_id: &AccountId,
+        proof: &[u8],
+        raw_tx: &[u8],
+        transaction: &Transaction,
+    ) -> Result<(), Error> {
         tracing::info!("Found tx from vault {}", vault_id.to_ss58check());
         // check if matching redeem or replace request
-        if self.btc_parachain.is_transaction_invalid(vault_id, &raw_tx).await? {
+        if self.btc_parachain.is_transaction_invalid(vault_id, raw_tx).await? {
             tracing::info!("Transaction is invalid");
             self.btc_parachain.report_vault_theft(vault_id, proof, raw_tx).await?;
+        } else {
+            // valid payment.. but check that it is not a duplicate payment
+            for (txid, block_hash) in self.bitcoin_core.find_duplicate_payments(transaction).await? {
+                let raw_tx_2 = self.bitcoin_core.get_raw_tx(&txid, &block_hash).await?;
+                let proof_2 = self.bitcoin_core.get_proof(txid, &block_hash).await?;
+                if !self.btc_parachain.is_transaction_invalid(vault_id, &raw_tx_2).await? {
```

Comment:
Both payments may be valid as the executed txid is not stored on chain. Remove this call, it's enough to see two op-return values from the same Vault.

Reply:
You are right, this call is not super necessary. To be clear, it checks that the transaction by itself _is_ valid. Because this is the goal of the double payment detection: to catch duplicate valid payments. If one of the transactions is _not_ valid, it is already reported by the regular theft check. I thought it a little bit cleaner to keep these two cases mutually exclusive

Reply:
You are correct, good point!